libraries/AC_Avoidance/AP_OABendyRuler.cpp

```diff
@@ -0,0 +1,294 @@
+const float OA_BENDYRULER_LOOKAHEAD_STEP2_RATIO = 1.0f; // step2's lookahead length as a ratio of step1's lookahead length
+const float OA_BENDYRULER_LOOKAHEAD_STEP2_MIN = 2.0f;   // step2 checks at least this many meters past step1's location
+const float OA_BENDYRULER_LOOKAHEAD_PAST_DEST = 2.0f;   // lookahead length will be at least this many meters past the destination
+const float OA_BENDYRULER_LOW_SPEED_SQUARED = 0.04f;    // when ground course is below this speed squared, vehicle's heading will be used (0.04 = 0.2m/s)
+
+// run background task to find best path and update avoidance_results
+// returns true and updates origin_new and destination_new if a best path has been found
+bool AP_OABendyRuler::update(const Location& current_loc, const Location& destination, const Vector2f &ground_speed_vec, Location &origin_new, Location &destination_new)
+{
+    // bendy ruler always sets origin to current_loc
+    origin_new = current_loc;
+
+    // calculate bearing and distance to final destination
+    const float bearing_to_dest = current_loc.get_bearing_to(destination) * 0.01f;
+    const float distance_to_dest = current_loc.get_distance(destination);
+
+    // lookahead distance is adjusted dynamically based on avoidance results
+    _current_lookahead = constrain_float(_current_lookahead, _lookahead * 0.5f, _lookahead);
+
+    // calculate lookahead dist and time for step1.  distance can be slightly longer than
+    // the distance to the destination to allow room to dodge after reaching the destination
+    const float lookahead_step1_dist = MIN(_current_lookahead, distance_to_dest + OA_BENDYRULER_LOOKAHEAD_PAST_DEST);
+
+    // calculate lookahead dist for step2
+    const float lookahead_step2_dist = _current_lookahead * OA_BENDYRULER_LOOKAHEAD_STEP2_RATIO;
+
+    // get ground course
+    float ground_course_deg;
+    if (ground_speed_vec.length_squared() < OA_BENDYRULER_LOW_SPEED_SQUARED) {
+        // with zero ground speed use vehicle's heading
+        ground_course_deg = AP::ahrs().yaw_sensor * 0.01f;
+    } else {
+        ground_course_deg = ground_speed_vec.angle();
+    }
+
+    // check OA_BEARING_INC definition allows checking in all directions
+    static_assert(360 % OA_BENDYRULER_BEARING_INC == 0, "check 360 is a multiple of OA_BEARING_INC");
+
+    // search in 5 degree increments around the vehicle alternating left
+    // and right. For each direction check if vehicle would avoid all obstacles
+    float best_bearing = bearing_to_dest;
+    bool have_best_bearing = false;
+    float best_margin = -FLT_MAX;
+    float best_margin_bearing = best_bearing;
+
+    for (uint16_t i = 0; i < (360 / OA_BENDYRULER_BEARING_INC); i++) {
+        float bearing_delta = i * OA_BENDYRULER_BEARING_INC * 0.5f;
+        if (i & 1) {
+            // we alternate between trying to the left and right of the target
+            bearing_delta = -bearing_delta;
```

Comment:
txs for finding this.  I can't imagine this is intentional so I've changed it to be two loops instead of one.  So the external loop will be from 0 to 170, the internal loop will be 0 to 1.  So it will look like this:

>for (uint8_t i = 0; i <= (170 / OA_BENDYRULER_BEARING_INC); i++) {
>        for (uint8_t bdir = 0; bdir <= 1; bdir++) {
>            // skip duplicate check of bearing straight towards destination
>            if ((i==0) && (bdir > 0)) {
>                continue;
>            }